### PR TITLE
6 send method variants

### PR DESCRIPTION
**Pull request by mauzigoe**
Proposed fix for Issue #6

### CODE REVIEW

**Comment by mauzigoe** [comment]
RequestExt has been split into 2 new Traits `RequestExt` and `RequestWithoutBodyExt`. The first one is implemented for `http::Request<T>` to use send() with body `http::Request<T>::body`, the second one `Request<()>` implements to send a body afterwards with send<B: IntoBodyRequest>(body: B).

At the moment RequestExt::send_with_client_configuration() implementation use a clone of self to extract the body. This is suboptimal and needs to be changed.

File: src/http/request_native.rs [hunk 1]
```diff
@@ -59,8 +59,8 @@ pub(crate) enum RequestSend<'a> {
 }
 
 impl RequestSend<'_> {
-    pub fn new_with_client_config<'a>(
-        request: &'a http::Request<()>,
+    pub fn new_with_client_config<'a,T: IntoRequestBody>(
```
Comment (FlorianUekermann) [inline src/http/request_native.rs]: Isn't the constraint on T unnecessary?

File: src/http/body.rs [hunk 2]
```diff
@@ -7,6 +7,13 @@ pub trait IntoRequestBody {
     fn into_request_body(self) -> (Self::RequestBody, u64);
 }
 
+pub trait NotEmptyBody: IntoRequestBody {}
```
Comment (FlorianUekermann) [inline src/http/body.rs]: Let's call this `IntoNonUnitRequestBody`

File: src/http/mod.rs [hunk 3]
```diff
@@ -126,3 +148,21 @@ impl Debug for ResponseBody {
         f.debug_struct("ResponseBody").finish()
     }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (FlorianUekermann) [inline src/http/mod.rs]: let's move this to body.rs and add one test below each trait impl, e.g.:
```rust
#[cfg(test)]
mod test_option {
  use crate::prelude::*;
  use http::Request;
  #[test]
  fn test_send() {
      Request::post("https://example.com").body(()).unwrap().send(Some(&[4u8]));
      Request::post("https://example.com").body(Some(&[4u8])).unwrap().send();
  }
}
```

File: src/http/mod.rs [hunk 4]
```diff
@@ -20,16 +21,37 @@ mod response_native;
...
+
+pub trait RequestWithoutBodyExt<'a>: Sized {
     #[cfg(any(feature = "ring", feature = "aws-lc-rs"))]
     fn send<B: IntoRequestBody + 'a>(self, body: B) -> RequestSend<'a> {
         let client_config = crate::DEFAULT_CLIENT_CONFIG.clone();
-        self.send_with_client_config(body, client_config)
+        self.send_with_body_and_client_config(body, client_config)
     }
-    fn send_with_client_config<B: IntoRequestBody + 'a>(self, body: B, client_config: Arc<ClientConfig>) -> RequestSend<'a>;
+    fn send_with_body_and_client_config<B: IntoRequestBody + 'a>(self, body: B, client_config: Arc<ClientConfig>) -> RequestSend<'a>;
```
Comment (FlorianUekermann) [inline src/http/mod.rs]: `send_with_client_config()`

File: src/http/mod.rs [hunk 5]
```diff
@@ -20,16 +21,37 @@ mod response_native;
 type RequestSendInner<'a> = request_native::RequestSend<'a>;
 
 pub trait RequestExt<'a>: Sized {
```
Comment (FlorianUekermann) [inline src/http/mod.rs]: `RequestWithBodyExt`

**Review by FlorianUekermann — CHANGES_REQUESTED** [review]
some suggestions, mostly naming consistency

prelude.rs requires update

try `cargo test --features websocket`

**Comment by FlorianUekermann** [comment]
> At the moment RequestExt::send_with_client_configuration() implementation use a clone of self to extract the body. This is suboptimal and needs to be changed.

Let's ignore that for now and deal with it in #10

File: examples/post.rs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-use async_web_client::prelude::*;
+use async_web_client::{prelude::*, RequestWithoutBodyExt};
```
Comment (FlorianUekermann) [inline examples/post.rs]: additional use shouldn't be necessary if prelude is complete

File: src/http/body.rs [hunk 7]
```diff
@@ -50,3 +57,24 @@ impl<T: IntoRequestBody> IntoRequestBody for Option<T> {
         }
     }
 }
+
```
Comment (FlorianUekermann) [inline src/http/body.rs:60]: We need one of these for each `impl IntoRequestBody for` like the example below for `Option<T>`.

```rust
#[cfg(test)]
mod test_option {
  use crate::prelude::*;
  use http::Request;
  #[test]
  fn test_send() {
      Request::post("https://example.com").body(()).unwrap().send(Some(&[4u8]));
      Request::post("https://example.com").body(Some(&[4u8])).unwrap().send();
  }
}
```

File: src/http/mod.rs [hunk 8]
```diff
@@ -28,7 +39,18 @@ pub trait RequestExt<'a>: Sized {
     fn send_with_client_config<B: IntoRequestBody + 'a>(self, body: B, client_config: Arc<ClientConfig>) -> RequestSend<'a>;
 }
 
-impl<'a> RequestExt<'a> for &'a http::Request<()> {
+impl<'a,T: IntoNonUnitRequestBody + Clone> RequestWithBodyExt<'a> for &'a http::Request<T> {
+    type B = T;
+    fn send_with_client_config(self, client_config: Arc<ClientConfig>) -> RequestSend<'a> {
+	let cloned_self = (*self).clone();
+        let (read, len) = cloned_self.into_body().into_request_body();
```
Comment (FlorianUekermann) [inline src/http/mod.rs:46]: doesn't this make the `clone()` unnecessary? (if it doesn't work, ignore it for now):
```rust
   let (head, body) = self.into_parts();
   let req = http::Request::from_parts(head, ());
   req.send_with_client_config(req, body, client_config)
```

Then we can also drop the additional type argument in `request_native.rs::RequestSend::new_with_client_config` and the `Clone` requirement here.

**Review by FlorianUekermann — CHANGES_REQUESTED** [review]

File: src/http/mod.rs (hunk 8 above)
Comment (mauzigoe) [reply src/http/mod.rs:46]: we implement `RequestWithbodyExt` for &'a Request<T>. Consuming `self` (when using `into_parts`) seems to be a problem.

**Pull request merged**